File: backend/src/main/java/com/marinelink/common/api/ApiResponse.java
package com.marinelink.common.api;

import com.fasterxml.jackson.annotation.JsonInclude;
import lombok.Builder;
import lombok.Getter;

import java.util.List;

/**
 * Unified API response envelope for all endpoints.
 *
 * <p>Success: {@code success=true}, {@code data} populated, {@code errors} null.
 * <p>Error: {@code success=false}, {@code data} null, {@code message} and {@code errors} populated.
 */
@Getter
@Builder
@JsonInclude(JsonInclude.Include.NON_NULL)
public class ApiResponse<T> {

    private final boolean success;
    private final String message;
    private final T data;
    private final List<FieldError> errors;
    private final PaginationMeta pagination;

    // ── Static factories ──────────────────────────────────────────────────────

    public static <T> ApiResponse<T> ok(T data) {
        return ApiResponse.<T>builder()
                .success(true)
                .message("OK")
                .data(data)
                .build();
    }

    public static <T> ApiResponse<T> ok(T data, String message) {
        return ApiResponse.<T>builder()
                .success(true)
                .message(message)
                .data(data)
                .build();
    }

    public static <T> ApiResponse<T> ok(T data, PaginationMeta pagination) {
        return ApiResponse.<T>builder()
                .success(true)
                .message("OK")
                .data(data)
                .pagination(pagination)
                .build();
    }

    public static <T> ApiResponse<T> created(T data, String message) {
        return ApiResponse.<T>builder()
                .success(true)
                .message(message)
                .data(data)
                .build();
    }

    public static <T> ApiResponse<T> error(String message) {
        return ApiResponse.<T>builder()
                .success(false)
                .message(message)
                .build();
    }

    public static <T> ApiResponse<T> error(String message, List<FieldError> errors) {
        return ApiResponse.<T>builder()
                .success(false)
                .message(message)
                .errors(errors)
                .build();
    }

    // ── Nested types ─────────────────────────────────────────────────────────

    @Getter
    @Builder
    public static class FieldError {
        private final String field;
        private final String message;
    }

    @Getter
    @Builder
    public static class PaginationMeta {
        private final int page;
        private final int size;
        private final long totalElements;
        private final int totalPages;

        public static PaginationMeta of(org.springframework.data.domain.Page<?> page) {
            return PaginationMeta.builder()
                    .page(page.getNumber())
                    .size(page.getSize())
                    .totalElements(page.getTotalElements())
                    .totalPages(page.getTotalPages())
                    .build();
        }
    }
}

File: frontend/lib/core/api/api_response.dart
/// Dart representation of the unified API response envelope.
///
/// Matches the JSON structure from MarineLink_API_Documentation.md:
/// ```json
/// { "success": true, "data": {}, "message": "OK", "pagination": {...} }
/// ```
class ApiResponse<T> {
  final bool success;
  final String? message;
  final T? data;
  final List<ApiFieldError>? errors;
  final ApiPagination? pagination;

  const ApiResponse({
    required this.success,
    this.message,
    this.data,
    this.errors,
    this.pagination,
  });

  factory ApiResponse.fromJson(
    Map<String, dynamic> json,
    T Function(dynamic json) fromJson,
  ) {
    return ApiResponse<T>(
      success: json['success'] as bool? ?? false,
      message: json['message'] as String?,
      data: json['data'] != null ? fromJson(json['data']) : null,
      errors: (json['errors'] as List<dynamic>?)
          ?.map((e) => ApiFieldError.fromJson(e as Map<String, dynamic>))
          .toList(),
      pagination: json['pagination'] != null
          ? ApiPagination.fromJson(json['pagination'] as Map<String, dynamic>)
          : null,
    );
  }
}

class ApiFieldError {
  final String? field;
  final String? message;

  const ApiFieldError({this.field, this.message});

  factory ApiFieldError.fromJson(Map<String, dynamic> json) {
    return ApiFieldError(
      field: json['field'] as String?,
      message: json['message'] as String?,
    );
  }
}

class ApiPagination {
  final int page;
  final int size;
  final int totalElements;
  final int totalPages;

  const ApiPagination({
    required this.page,
    required this.size,
    required this.totalElements,
    required this.totalPages,
  });

  factory ApiPagination.fromJson(Map<String, dynamic> json) {
    return ApiPagination(
      page: json['page'] as int? ?? 0,
      size: json['size'] as int? ?? 20,
      totalElements: json['totalElements'] as int? ?? 0,
      totalPages: json['totalPages'] as int? ?? 0,
    );
  }
}
